Add New Spectrograph: LDT/DeVeny

tbowers7 opened the pull request:
Introduce the Lowell Discovery Telescope / DeVeny Optical Spectrograph to the PypeIt repository.  This is a long-slit spectrograph, formerly known as the KPNO White Spectrograph (used on the No.1 36-inch Telescope), on long-term loan to Lowell Observatory.  It is currently mounted on one of the cassegrain ports of the 4.3-m LDT near Flagstaff, AZ.

This PR includes the new spectrograph class along with telescope parameters for the LDT.  A reid_arxiv template has been added for one of our gratings for which `holy-grail` consistently fails.  Furthermore, a consistently identified cadmium line (6440.25Å) has been added to that line list.

Documentation has been updated in all of the locations mentioned at https://pypeit.readthedocs.io/en/latest/dev/development.html .

tbowers7 committed:
Add CdI line at 6440.2490Å from NIST line list

The CdI line at 6440.2490Å is easily detected in LDT/DeVeny spectra.  Added
this line by hand from the NIST line list for neutral cadmium.  Vacuum
wavelength found in the CdI_vacuum.ascii file in the
pypeit/data/arc_lines/NIST directory.

Note: Intensity entered in this commit not actual value, but is merely a
placeholder.

Line addition approved by J. Xavier Prochaska in Slack #wavecalib on 2021-06-17.

	modified:   pypeit/data/arc_lines/lists/CdI_lines.dat

tbowers7 committed:
Add LDT / DeVeny Spectrograph class

Following the example of the Bok/B&C Spectrograph addition, add parameters
for the Lowell Discovery Telescope / DeVeny Optical Spectrograph.  This is a
long-slit spectrograph, formerly known as the KPNO White Spectrograph (used on
the No.1 36-inch Telescope), on long-term loan to Lowell Observatory.  It is
currently mounted on one of the cassegrain ports of the 4.3-m LDT near
Flagstaff, AZ.

First pass at the class LDTDeVenySpectrograph() for addition to the PypeIt
repository.  Add the LDT/DeVeny to parameter, telescope, and init files, as
well as front-facing spectrograph lists.

This commit represents a thorough review of all PypeIt parameters, and making
adjustments based on the specific needs of the LDT/DeVeny spectrograph.

Overview:
For all calibration frames, use median combine rather than weighted mean
Specify the arc lamps available on DeVeny
Tweaks to some wavelength calibration values
Add fix for PypeIt issue #1155 (ech_fix_format = False)
Slit-edge setting for long-slit spectrographs
A couple default exposure times (bias/arc), otherwise there is no 'normal' range
A couple extraction and flux calibration adjustments

The parameters are now in the same order as shown in the online PypeIt
documentation for ease of reference.

	modified:   README.md
	modified:   doc/index.rst
	modified:   pypeit/par/pypeitpar.py
	modified:   pypeit/spectrographs/__init__.py
	new file:   pypeit/spectrographs/ldt_deveny.py
	modified:   pypeit/telescopes.py

tbowers7 committed:
Separate 'science' and 'standard' frametypes for LDT/DeVeny

Split out 'science' and 'standard' frametypes instead of lumping them together.

Many of the DeVeny users do not care about spectrophotometric standards because
flux calibration is not relevant for their science.  Therefore, having to remove
'standard' from all science frames in the PypeIt Reduction File would be a bit
cumbersome.

We are exploring adding a FITS IMAGETYP/OBSTYPE = 'STANDARD' for observations
specifically of spectrophotometric standard stars.  Then, for those DeVeny users
who take such frames, PypeIt will recognize those as standards without the user
needing to adjust the frametype manually.

	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Add DeVeny dark current measurement

	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Add reid_arxiv template for DV5 grating

Add a reid_arxiv template for the DV5 grating, constructed using the
`templates.build_template()` function from three different spectra spanning the
useful spectral range for this grating.

This one grating seems to confound the `holy-grail` wavelength calibration
method, so this reid_arxiv is used with `full-template` for the DV5 (500/5500)
grating for successful calibration.

	new file:   pypeit/data/arc_lines/reid_arxiv/ldt_deveny_500l_HgCdAr.fits

tbowers7 committed:
Add full set of telescope paramters for LDT

In `LDTTelescopePar()`, add `fratio`, `diameter`, and `eff_aperture` parameters
for the Lowell Discovery Telescope.

	modified:   pypeit/telescopes.py

tbowers7 committed:
LDT/DeVeny supported = True

	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Add unit test to tickle I/O for LDT/DeVeny

	modified:   pypeit/tests/test_load_images.py

tbowers7 committed:
Remove all exposure time ranges for frame typing

The FITS IMAGETYP --> PypeIt frametype conversions in `check_frame_type()` are
sufficient for identifying frames, and exposure time based checks are not
necessary for LDT/DeVeny data.

	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Add new Cooked file to test_cooked.py

Add v1.4.5 in preparation of adding LDT/DeVeny to the develop branch.

	modified:   pypeit/tests/test_cooked.py

tbowers7 committed:
Add these changes to CHANGES.rst

	modified:   CHANGES.rst

tbowers7 committed:
Doc updates for PR

	new file:   doc/api/pypeit.spectrographs.ldt_deveny.rst
	modified:   doc/api/pypeit.spectrographs.rst
	modified:   doc/help/pypeit_chk_for_calibs.rst
	modified:   doc/help/pypeit_obslog.rst
	modified:   doc/help/pypeit_setup.rst
	modified:   doc/help/pypeit_trace_edges.rst
	modified:   doc/help/pypeit_view_fits.rst
	modified:   doc/help/run_pypeit.rst
	modified:   doc/include/spectrographs_table.rst
	modified:   doc/pypeit_par.rst

tbowers7 committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into add_ldt_deveny

tbowers7 committed:
Fix spurious documentation changes

	modified:   doc/help/pypeit_obslog.rst
	modified:   doc/help/pypeit_setup.rst
	modified:   doc/help/run_pypeit.rst
	modified:   doc/pypeit_par.rst

tbowers7 commented:
Simultaneous PR to the development suite.

Test results:
```
--- PYPEIT DEVELOPMENT SUITE FAILED 1/120 TESTS (Masters ignored) ---
Failed tests:
    bok_bc/600 pypeit (without masters)
Skipped tests:
```

Not sure why the `bok/bc` tests failed, but all LDT/DeVeny tests passed.

Cooked uploaded as `Cooked_pypeit_dev_v1.4.5.tar.gz`

Two tests failed, but these are not related to the new LDT/DeVeny code:
```
test_alignments.py .                                [  0%]
test_arc.py .                                       [  0%]
test_archive.py ..                                  [  1%]
test_arcimage.py ..                                 [  2%]
test_biasframe.py ..                                [  2%]
test_bitmask.py ....                                [  4%]
test_bpmimage.py ..                                 [  4%]
test_bspline.py ..........                          [  8%]
test_buildimage.py .                                [  8%]
test_calibrations.py .......                        [ 11%]
test_chk_calibs.py ..                               [ 12%]
test_coadd.py .                                     [ 12%]
test_collate_1d.py ..........                       [ 16%]
test_cooked.py .                                    [ 16%]
test_coremeta.py .                                  [ 16%]
test_datacontainer.py ........                      [ 19%]
test_detector.py ...                                [ 20%]
test_display.py .                                   [ 20%]
test_edgetrace.py .                                 [ 21%]
test_fitting.py ....                                [ 22%]
test_flatfield.py .                                 [ 23%]
test_flexure.py ..                                  [ 23%]
test_flux.py ....                                   [ 25%]
test_fluxspec.py ..s                                [ 26%]
test_frametype.py .                                 [ 26%]
test_history.py ....                                [ 28%]
test_load.py .                                      [ 28%]
test_load_images.py ................                [ 34%]
test_maskdesign_matching.py ...                     [ 35%]
test_masterframe.py .                               [ 35%]
test_match.py ....                                  [ 37%]
test_metadata.py ......                             [ 39%]
test_moment.py ......                               [ 41%]
test_msgs.py ..                                     [ 41%]
test_opticalmodel.py ....                           [ 43%]
test_parse.py ...                                   [ 44%]
test_pca.py ........                                [ 47%]
test_processimage.py ..                             [ 48%]
test_procimg.py .                                   [ 48%]
test_pydl.py .                                      [ 48%]
test_pypeit.py ..                                   [ 49%]
test_pypeitimage.py .                               [ 49%]
test_pypeitpar.py ................................  [ 61%]
test_pypeitsetup.py .......                         [ 63%]
test_qa.py .                                        [ 64%]
test_runpypeit.py ..                                [ 64%]
test_scienceimage.py ...                            [ 65%]
test_scripts.py ...F..........                      [ 70%]
test_setups.py ....................F                [ 78%]
test_skysub.py ..                                   [ 79%]
test_slitmask.py ...                                [ 80%]
test_slittrace.py ....                              [ 81%]
test_spec2dobj.py ......                            [ 83%]
test_specobj.py .......                             [ 86%]
test_specobjs.py .....                              [ 87%]
test_spectrographs.py ....................          [ 95%]
test_syncspec.py .                                  [ 95%]
test_traceimage.py .                                [ 95%]
test_utils.py ...                                   [ 96%]
test_vcorr.py ..                                    [ 97%]
test_wavemodel.py .                                 [ 97%]
test_wavetilts.py ....                              [ 99%]
test_wvcalib.py ..                                  [100%]

=========================== short test summary info ===========================
FAILED test_scripts.py::test_show_1dspec
FAILED test_setups.py::test_setup_vlt_fors2
===== 2 failed, 278 passed, 1 skipped, 4619 warnings in 1488.39s (0:24:48) ====
```

profxj commented on pypeit/spectrographs/ldt_deveny.py:
```diff
@@ -0,0 +1,399 @@
+            gain            = np.atleast_1d(header['GAIN']),
+            ronoise         = np.atleast_1d(header['RDNOISE']),
+            # Data & Overscan Sections -- Edge tracing can handle slit edges
+            datasec         = np.atleast_1d('[5:512,54:2096]'),
```
Just checking, but can observers modify the 'windowing'?

If so, they will someday and his will break..

tbowers7 replied:
In theory, an adventurous observer could try to define a subframe, but this is a vestigial function leftover in the interface from the development of another instrument, is not included in any of our user documentation, and would not gain the observer much time since the chip reads out fairly quickly as it is.  Do you have a suggestion for a different method for defining the `datasec` and `oscansec`?  The related FITS keywords (TRIMSEC and BIASSEC) are hardwired at these values with our present controller software.

profxj replied:
This is fine then.  Maybe add a brief comment in the
code summarizing what you wrote above.  Thanks.

profxj commented on pypeit/spectrographs/ldt_deveny.py:
```diff
@@ -0,0 +1,399 @@
+        par['calibrations']['pixelflatframe']['process']['combine'] = 'median'
+        par['calibrations']['illumflatframe']['process']['combine'] = 'median'
+
+        # For science, standard, and sky frames, combine using weighted mean (Default)
```
remove

tbowers7 replied:
Done.

profxj commented on pypeit/spectrographs/ldt_deveny.py:
```diff
@@ -0,0 +1,399 @@
+        set_use = dict(use_illumflat=False)
+        par.reset_all_processimages_par(**set_use)
+
+        # Use median combine (rather than weighted mean) for all cals
```
maybe explain why

tbowers7 replied:
Hrm, "because I've always just done it that way" is not a very good justification.  Upon careful consideration, changed all calibration combines except `pixelflatframe` to `weightmean`.  Added brief explanation to comment in code.

profxj commented on pypeit/spectrographs/ldt_deveny.py:
```diff
@@ -0,0 +1,399 @@
+        #  the wavelength calibration method `reidentify`.
+        par['calibrations']['wavelengths']['ech_fix_format'] = False
+
+        # Slit-edge settings for long-slit data (slit > 90" long)
```
Is this the only slit allowed?  Just checking.

tbowers7 replied:
The spectrograph has but one slit that nearly completely illuminates the CCD.  There are reflection features in the data at the slit edges (reflective surfaces near the slit jaws) that would frequently cause the edge tracing algorithm to identify narrow slits at the edge of the CCD rather than the actual slit.  The parameters in lines 228 - 230 were those found to allow the edge tracing algorithm to consistently correctly identify the slit.

There are plans to install a decker at some point in the future, and when the opening size(s) are determined, this parameter will be appropriately modified.

profxj commented on pypeit/spectrographs/ldt_deveny.py:
```diff
@@ -0,0 +1,399 @@
+        # Call the base-class method to generate the empty bpm
+        bpm_img = super().bpm(filename, det, shape=shape, msbias=msbias)
+
+        # msgs.info("Using hard-coded BPM for DeVeny")
```
Are there any bad columns on the detector?  If not, this is fine.
If so, I'd suggest masking them..

tbowers7 replied:
No bad columns.

profxj replied:
I think you can therefore remove this method and use the default one.

profxj commented on pypeit/tests/test_cooked.py:
```diff
@@ -20,6 +20,6 @@ def test_cooked_version():
     with open(v_file) as f:
         tmp = f.readlines()
-    assert tmp[-1].strip() == '1.4.4'
+    assert tmp[-1].strip() == '1.4.5'
```
I see you uploaded a new Cooked file.  Did you add files to it?
I'm guessing not given what I've seen (i.e. no new tests other
than the `load_image`.  

If no changes, you can skip that step, remove the 1.4.5. file and set this back.

tbowers7 replied:
Reverted.

profxj commented on pypeit/tests/test_load_images.py:
```diff
@@ -183,6 +183,15 @@ def test_load_goodman():
     except:
         pytest.fail('Bok BC test data section failed: {0}'.format(ifile))
 
+@dev_suite_required
```
Nice!

tbowers7 replied:
Are there other tests I should add as part of integrating LDT/DeVeny into PypeIt?

profxj replied:
Nah.  Thanks for asking!

profxj reviewed (changes requested):
This looks great!

I've put in mainly clarification questions as I am 
not familiar with the instrument.  

Thanks!!

profxj commented:
Pasting from the DevSuite PR:

--- PYPEIT DEVELOPMENT SUITE FAILED 1/120 TESTS (Masters ignored) ---
Failed tests:
    bok_bc/600 pypeit (without masters)
Skipped tests:

tbowers7 committed:
Revert Cooked data test

	modified:   pypeit/tests/test_cooked.py

tbowers7 committed:
Adjust image combination methods for calibrations

Because of the intrinsic shape and structure of flat field frames, doing image
combination using average with sigma clipping doesn't work so well.  Leave this
frame type with `median` combine.  Use `weightmean` for all other frame types
since this will yield marginally better noise properties.

	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into add_ldt_deveny

tbowers7 committed:
Modify doc to reflect change in image combination parameters

	modified:   doc/pypeit_par.rst

tbowers7 committed:
Merge branch 'develop' into add_ldt_deveny

tbowers7 committed:
Merge branch 'develop' of https://github.com/pypeit/PypeIt into add_ldt_deveny

tbowers7 committed:
Responding to change requests in PR

Add comments in the code as to why the `datasec` and `oscansec` values are
hardwired.

Remvoved the bad pixel mask method because there are currently no bad columns
on the CCD.  If this were to change in the future, this method would need to
be reinserted.

	modified:   pypeit/spectrographs/ldt_deveny.py

kbwestfall commented on pypeit/data/arc_lines/lists/CdI_lines.dat:
```diff
@@ -1,3 +1,4 @@
+# Creation Date: 2021-Jun-17
```
Remove the line below that gives a different creation date.

tbowers7 replied:
Done.  I was following other linelists that showed multiple creation dates (e.g., ArI, HgI, CuI, etc.).

profxj reviewed (approved):
approving

kbwestfall commented on pypeit/spectrographs/ldt_deveny.py:
```diff
@@ -0,0 +1,357 @@
+            ids = list(gratings.keys())
+            kwds = list(gratings.values())
+            idx = kwds.index(headarr[0]['GRATING'])
+            return f"{ids[idx]} ({kwds[idx]})"
```
No need to change anything, but this feels round about.  You define the dictionary as key = new name and value = old name, but you want to select based on the old name.  Why not define the dictionary in opposite way so that you return something like `f"{gratings[headarr[0]['GRATING']]} {headarr[0]['GRATING']}"` ?  Regardless, you might want to put in a check that the header keyword is in the list of possibilities, and raise an exception if not.

tbowers7 replied:
Check added, although the GRATING FITS keyword is limited to a hardwired list of choices, with `UNKNOWN` being inserted if one of the options is not selected.  The dictionary has been flipped, as suggested.

kbwestfall commented on pypeit/spectrographs/ldt_deveny.py:
```diff
@@ -0,0 +1,357 @@
+                decker = headarr[0]['DECKER']
+            else:
+                decker = "None"
+            return decker
```
`return headarr[0]['DECKER'] if 'DECKER' in headarr[0].keys() else 'None'` ?

tbowers7 replied:
This is a stub for when we add a decker plate within the next year.  There is currently no FITS keyword for the decker, but there will be when it is installed.

tbowers7 replied:
Elegant coding is not my strong suit.  Thank you for the suggestion.

kbwestfall reviewed (approved):
Looks great!  Thanks for this very complete integration of this new spectrograph into `PypeIt`!  I just had a couple nit-picky code-style comments.  @profxj , I'm happy for you to merge when you are.

kbwestfall commented:
Hi @tbowers7 .  Actually, one more thing:  Originally you posted that 2 of the unit tests were failing.  Is that still true?  Even if the code being tested doesn't seem like it touches the code you edited, we need all the unit tests to pass before we merge this branch.  Same thing goes for the dev-suite tests.  Let me know if they continue to fail, and I can try to help you debug them.

tbowers7 commented:
I will rerun the tests today.

tbowers7 committed:
Addressing PR Change Requests

	modified:   pypeit/data/arc_lines/lists/CdI_lines.dat
	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 commented:
@kbwestfall The two unit tests are still failing (still re-running the dev suite):
```
=========================== short test summary info ===========================
FAILED test_scripts.py::test_show_1dspec
FAILED test_setups.py::test_setup_vlt_fors2
```
Here are more detailed outputs:
```
_______________________________ test_show_1dspec _______________________________
.
.
.
if API in PYSIDE_API:
        try:
>           from PySide import __version__ as PYSIDE_VERSION  # analysis:ignore
E           ModuleNotFoundError: No module named 'PySide'

/home/tbowers/d1/codes/PypeIt/.tox/test/lib/python3.9/site-packages/qtpy/__init__.py:204: ModuleNotFoundError

During handling of the above exception, another exception occurred:
.
.
.
if API in PYSIDE_API:
        try:
            from PySide import __version__ as PYSIDE_VERSION  # analysis:ignore
            from PySide.QtCore import __version__ as QT_VERSION  # analysis:ignore
            PYQT_VERSION = None
            PYQT5 = PYSIDE2 = False
            PYSIDE = True
        except ImportError:
>           raise PythonQtError('No Qt bindings could be found')
E           qtpy.PythonQtError: No Qt bindings could be found

/home/tbowers/d1/codes/PypeIt/.tox/test/lib/python3.9/site-packages/qtpy/__init__.py:210: PythonQtError
```
And
```
_____________________________ test_setup_vlt_fors2 _____________________________
.
.
.
# Setup the table. Need to use object type for strings so that
        # they're not truncated.
        answers = table.Table()
        answers['setups'] = list(uniq_cfg.keys())
        # Add the configuration columns
        for setup, setup_dict in uniq_cfg.items():
            for key, value in setup_dict.items():
                answers[key] = np.empty(len(answers), dtype=object) if isinstance(value, str) \
>                                   else type(value)(0)
E               TypeError: NoneType takes no arguments

/home/tbowers/d1/codes/PypeIt/.tox/test/lib/python3.9/site-packages/pypeit/scripts/chk_for_calibs.py:90: TypeError
```

The `test_show_1dspec` error may be because I cannot seem to install `PySide` (not `PySide2`), but I do have the `pyqt5` package installed, so not sure why the code is searching for `PySide` in the first place.

Not really sure what's going on with the `test_setup_vlt_fors2` error.

profxj commented:
I can't see any reason for the VLT error myself, but can confirm it passes for me on `develop`:

(base) profx> py.test test_setups.py::test_setup_vlt_fors2
============================= test session starts =============================
platform linux -- Python 3.8.5, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /data/Projects/Python/PypeIt, configfile: setup.cfg
plugins: openfiles-0.5.0, ligo.skymap-0.5.0, filter-subpackage-0.1.1, hypothesis-6.8.5, astropy-header-0.1.2, cov-2.11.1, doctestplus-0.9.0, arraydiff-0.3, remotedata-0.3.2
collected 1 item                                                              

test_setups.py .                                                        [100%]

============================== warnings summary ===============================
../core/flux_calib.py:985
   DeprecationWarning: invalid escape sequence \e (flux_calib.py:985)

../core/flux_calib.py:1009
   DeprecationWarning: invalid escape sequence \e (flux_calib.py:1009)

../../../../../../home/xavier/Projects/anaconda3/lib/python3.8/site-packages/pkg_resources/__init__.py:1143
../../../../../../home/xavier/Projects/anaconda3/lib/python3.8/site-packages/pkg_resources/__init__.py:1143
../../../../../../home/xavier/Projects/anaconda3/lib/python3.8/site-packages/pkg_resources/__init__.py:1143
   DeprecationWarning: Use of .. or absolute path in a resource path is not allowed and will raise exceptions in a future release. (__init__.py:1143)

../sensfunc.py:40
   DeprecationWarning: invalid escape sequence \l (sensfunc.py:40)

../sensfunc.py:699
   DeprecationWarning: invalid escape sequence \l (sensfunc.py:699)

../sensfunc.py:822
   DeprecationWarning: invalid escape sequence \l (sensfunc.py:822)

../core/extract.py:2238
   DeprecationWarning: invalid escape sequence \{ (extract.py:2238)

pypeit/tests/test_setups.py::test_setup_vlt_fors2
pypeit/tests/test_setups.py::test_setup_vlt_fors2
pypeit/tests/test_setups.py::test_setup_vlt_fors2
pypeit/tests/test_setups.py::test_setup_vlt_fors2
pypeit/tests/test_setups.py::test_setup_vlt_fors2
   UserWarning: No parameters necessary for median overscan method.  Ignoring input. (pypeitpar.py:443)

-- Docs: https://docs.pytest.org/en/stable/warnings.html
======================= 1 passed, 14 warnings in 4.06s ========================

profxj commented:
And I'm willing to let the `PYSIDE` issue go if @finagle29 sees no harm

tbowers7 commented:
@kbwestfall Bok/B&C now passing the dev-suite (don't know what's different from last week):
```
Running tests on the following instruments:
    bok_bc

Reducing data from bok_bc for the following setups:
    600

 0 passed/ 0 failed/ 0 skipped STARTED bok_bc/600 pypeit (without masters)
 1 passed/ 0 failed/ 0 skipped PASSED  bok_bc/600 pypeit (without masters)

--- PYPEIT DEVELOPMENT SUITE PASSED 1/1 TESTS (Masters ignored) ---
Testing Started at 2021-06-30T11:33:26.917242
Testing Completed at 2021-06-30T11:35:12.039341
Total Time: 0:01:45.122099
```

Still running the full suite (it should be done late tonight), if you need to see a 100% passing rate.

profxj commented:
merging and will run tests on `develop` momentarily

tbowers7 commented:
Thank you!
